debug exchanges

buck54321 opened the pull request:
This PR is for testing. The exchanges package seems to have some issues lately that may or may not be related to recent work to implement websockets. This PR effectively just adds some trace output, though the foundations are being laid for some more sophisticated error monitoring. 

If you are testing, run this PR with `--debuglevel=XBOT=trace` on dcrdata, and `--loglevel=trace` on rateserver. Feel free to give me your logs.

chappjc commented on exchanges/exchanges_test.go:
```diff
@@ -283,7 +298,11 @@ func (p *fakePoloniexWebsocket) Write(interface{}) error {
+	select {
+	case <-poloniexDoneChannel:
+	default:
+		close(poloniexDoneChannel)
```
Is Close() possibly called in different places?  I'm trying to understand why this select is needed.  But it is a race since the channel could be closed by another goroutine if Close() is called elsewhere (they could both end up at the `close` statement and one would panic)

buck54321 replied:
I think it's unlikely during the test, but easy to add some sequencing.

chappjc commented on exchanges/websocket.go:
```diff
@@ -77,7 +77,14 @@ func (client *socketClient) Done() chan struct{} {
+	select {
+	case <-client.done:
+	default:
+		close(client.done)
```
Same comment as with the select in the test file.

chappjc replied:
BTW, `IsOpen` does not race per se, but the caller is likely racing.

buck54321 commented on exchanges/exchanges.go:
```diff
@@ -614,29 +610,21 @@ func (xc *CommonExchange) wsSend(msg interface{}) error {
-	return ws.On()
+	xc.wsMtx.RLock()
+	defer xc.wsMtx.RUnlock()
+	return xc.wsSync.init.After(xc.wsSync.fail)
```
Should have been this way already.

buck54321 commented on exchanges/exchanges_test.go:
```diff
@@ -27,157 +23,6 @@ func enableTestLog() {
-	enableTestLog()
-
-	// Skip this test during automated testing.
-	if os.Getenv("GORACE") != "" {
```
This was a sloppy way to skip this test during CI testing, but it's not compatible with the changes in #1347.

chappjc replied:
Is it because the test takes too much time that you wanted to skip it?  We can leverage the `-short` switch.  https://golang.org/pkg/testing/#hdr-Skipping e.g.

```go
if !testing.Short() {
    // do long tests
...
}
```

chappjc replied:
Or use `t.Skip` above I suppose.

buck54321 replied:
Yeah. All the live tests take a long time. I put a tag at the top of the _live file, so you have to enable it e.g. `go test -tags-livexc`.

chappjc replied:
Oh, perfect.

buck54321 commented:
@chappjc I think I've addressed the concerns you've raised. Before I remove the draft status, could you check out those areas and give me a thumbs up or down.

chappjc commented:
Great, thanks! I'll go though it this evening.
